File: Labs/Lab 8/lab8.java
import java.util.Arrays;

public class lab8 {
    public int coinChange(int[] coins, int amount) {
        //dp array, stores the min # of coins needed for each amount
        int[] dp = new int[amount + 1];
        
        //Fils the dp array w/ a default value which is higher than the maximum possible value
        Arrays.fill(dp, amount + 1);
        
        //base case
        dp[0] = 0;

        for (int a = 1; a <= amount; a++) {
            //see if each coin can contribute to the amount
            for (int c : coins) {
                //check if coin is not larger than the amount 
                if (a - c >= 0) {
                    // Set the dp at index a to the minimum of its current value,
                    // or the value of dp[a - c] + 1 (1 represents using the coin c)
                    dp[a] = Math.min(dp[a], dp[a - c] + 1);
                }
            }
        }
        
        //If dp[amount] is still set to the default value, we return -1 as it's not possible to make change
        return dp[amount] > amount ? -1 : dp[amount];
    }
}

File: Labs/Lab 8/lab8.py
from ast import List

class lab8:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount +1 ] * (amount +1) 
        dp[0] = 0 # base case

        for a in range(1, amount +1):
            for c in coins:
                if a - c >= 0:
                    dp[a] = min(dp[a], 1 + dp[a-c])

        return dp[amount] if dp[amount] != amount +1 else -1
    #if we cant compute it return -1